File: problems_0-1000/283-Move-Zeroes/python/solution.py
class Solution:
    def moveZeroes(self, nums: list[int]) -> None:
        nonZeroPtr = 0
        currentPtr = 0
        while currentPtr < len(nums):
            if nums[currentPtr] != 0:
                nums[currentPtr], nums[nonZeroPtr] = nums[nonZeroPtr], nums[currentPtr] 
                nonZeroPtr += 1
            currentPtr += 1
    
solution = Solution()
nums = [0, 1, 0, 3, 12]
solution.moveZeroes(nums)
print(nums)

File: problems_0-1000/283-Move-Zeroes/java/Solution.java
public class Solution {

    public void moveZeroes(int[] nums) {
        int nonZeroPtr = 0;
        int currentPtr = 0;
        while (currentPtr < nums.length) {
            if (nums[currentPtr] != 0) {
                int temp = nums[nonZeroPtr];
                nums[nonZeroPtr] = nums[currentPtr];
                nums[currentPtr] = temp;
                nonZeroPtr++;
            }
            
            currentPtr++;            
        }
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        int[] nums = new int[]{0, 1, 0, 3, 12};
        solution.moveZeroes(nums);
        for (int num : nums) {
            System.out.print(num + " ");
        }
    }
}
